src/main/java/uk/ac/ebi/pride/spectracluster/engine/ReassessingIncrementalClusteringEngine.java

```diff
@@ -0,0 +1,432 @@
+        this.similarityChecker = sck;
+        this.spectrumComparator = scm;
+        this.windowSize = windowSize;
+        this.similarityThreshold = similarityThreshold;
+        this.similarityRetainThreshold = DEFAULT_RETAIN_THRESHOLD;
+    }
+
+    public ReassessingIncrementalClusteringEngine(ISimilarityChecker similarityChecker, Comparator<ICluster> spectrumComparator, double windowSize, double similarityThreshold, double similarityRetainThreshold) {
+        this.similarityChecker = similarityChecker;
+        this.spectrumComparator = spectrumComparator;
+        this.windowSize = windowSize;
+        this.similarityThreshold = similarityThreshold;
+        this.similarityRetainThreshold = similarityRetainThreshold;
+    }
+
+    public double getWindowSize() {
+        return windowSize;
+    }
+
+
+    public int getCurrentMZ() {
+        return currentMZAsInt;
+    }
+
+
+    public void setCurrentMZ(final double pCurrentMZ) {
+        int test = MZIntensityUtilities.mzToInt(pCurrentMZ);
+        final int currentMZ = getCurrentMZ();
+        if (currentMZ > test) {  // all ow roundoff error but not much
+            double del = currentMZ - test;  // difference
+
+            if (Math.abs(del) > MZIntensityUtilities.MZ_RESOLUTION * MZIntensityUtilities.SMALL_MZ_DIFFERENCE) {
+                throw new IllegalStateException("mz values MUST be added in order - was "
+                        + NumberUtilities.formatDouble(currentMZ, 3) + " new " +
+                        NumberUtilities.formatDouble(pCurrentMZ, 3) + " del  " +
+                        del
+                );
+
+            }
+
+        }
+        currentMZAsInt = test;
+    }
+
+
+    /**
+     * Get clustered clusters
+     */
+    @Override
+    public List<ICluster> getClusters() {
+        final ArrayList<ICluster> ret = new ArrayList<ICluster>(clusters);
+        Collections.sort(ret);
+        return ret;
+    }
+
+    /**
+     * add some clusters
+     */
+    @Override
+    public void addClusters(ICluster... cluster) {
+        // or use a WrappedIncrementalClusteringEngine
+        throw new UnsupportedOperationException("Use addClusterIncremental instead or use a WrappedIncrementalClusteringEngine ");
+
+    }
+
+    /**
+     * add one cluster and return any clusters which are too far in mz from further consideration
+     *
+     * @return !null Cluster
+     */
+    @Override
+    public List<ICluster> addClusterIncremental(final ICluster added) {
+        double precursorMz = added.getPrecursorMz();
+        List<ICluster> clustersToremove = findClustersTooLow(precursorMz);
+        // either add as an existing cluster if make a new cluster
+        addToClusters(added);
+
+        // remove any non-fitting spectra
+        List<ICluster> nonFittingSpectra = removeNoneFittingSpectra();
+
+        // try to add them again by merging all clusters
+        internalGetClusters().addAll(nonFittingSpectra);
+        mergeAllClusters();
```

Comment:
The logic here should probably be arranged in a different order. mergeAllClusters will produce dirty clusters. When the next time addClusterINcremental is called, findClustersTooLow is called first, which will emit clustersToRemove that could contain dirty clusters as well.